lib/fluent/plugin/out_elasticsearch.rb

```diff
@@ -319,17 +330,19 @@ def write_objects(tag, chunk)
     meta = {}
     @error = Fluent::ElasticsearchErrorHandler.new(self)
 
+    records = []
     chunk.msgpack_each do |time, record|
       @error.records += 1
       next unless record.is_a? Hash
       begin
         process_message(tag, meta, header, time, record, bulk_message)
+        records << { time: time, record: record }
```

Comment:
Won't this use more memory on the non-error path as well, which we hope would be the fast path?  Is there a way we can just pass the "chunk" object and only when we find some errors, we index into the chunk object to pull out the original records to re-submit?

In general, it seems that the non-error case should not have to shoulder any of the burden for error handling, and the error path becomes more costly to compensate.

Reply:
in fact, you have to use the raw record from the chunk - you need the resubmitted record to look exactly like the raw record from the chunk, so it can be reprocessed.  So yes - we need the chunk, and in the error handing code we'll need to iterate the records using `chunk.msgpack_each do |time, record|`

Reply:
Agree with your observation but Rich is correct, we are basing the functionality on the assumption that the order of records submitted is the same as the items in the response.  This assumption is based upon Elasticsearch documentation as there is no way to otherwise correlate the request and response without ids or some hashing.  This is also necessary to only keep a list of the records that can be properly deserialized

Reply:
But `records` ends up having the records in the same order as `chunk.msgpack_each` emits them, right?  If that is the case, why do we need to construct a separate array on the normal path?  Why not pass chunk in, and when we encounter errors, then construct the list to handle the errors?

Reply:
We have to iterate the chunk in the error handling.  We need to make sure to get the same records that were sent, which means adding a couple of the record processing steps: omit records that are not Hash objects, and omit records that were sent to the dlq i.e. not included in the bulk index request.  We can't do any more processing on the records - records resubmitted to fluentd must be the raw records directly from the queue.

Reply:
Again, we cant iterate the chunk but the list of records that were actually submitted.